feat(move): Add `override_album_path_config` hook

ali-ramadhan opened the pull request:
- [x] I have read the contributing guide in the documentation.

### Description
This PR adds a new hook to the move plugin called `override_album_path_config` which allows plugins to modify the album path configuration dynamically based on the album's properties.

For more context, see discussion #231.

### Additional information

@jtpavlock I took a stab at implementing the `override_album_path_config` hook we discussed and added some tests which pass locally! But I'm not totally sure that I adhered to moe's design/coding philosophy as I'm still familiarizing myself with the code base. Would appreciate any feedback!

<!-- readthedocs-preview mrmoe start -->
----
📚 Documentation preview 📚: https://mrmoe--281.org.readthedocs.build/en/281/

<!-- readthedocs-preview mrmoe end -->

ali-ramadhan commented:
Ah woops I forgot to follow the commit message guidelines. I should be able to rewrite them on this branch.

jtpavlock commented on moe/move/move_core.py:
```diff
@@ -37,6 +37,22 @@ def create_path_template_func() -> list[Callable]:  # type: ignore
+        """Override the album path configuration.
+
+        This hook allows plugins to modify the album path configuration dynamically
+        based on the album's properties.
```
To me, it isn't quite clear how the new hook differentiates from `create_path_template_func` based on this description. 

Maybe "allows plugins to programmatically override the user's album path configuration"?

I also think an example would also be useful. The syntax for an example in a docstring is:

```python
        Example:
            .. code:: python

                settings.validators.register(
                    dynaconf.Validator("MOVE.ASCIIFY_PATHS", must_exist=True)
                )
```

The classical/soundtrack function you use in tests may be a perfect example.

ali-ramadhan replied:
Thanks for the syntax! I've improved the docstring and added an example.

jtpavlock commented on moe/move/move_core.py:
```diff
@@ -37,6 +37,22 @@ def create_path_template_func() -> list[Callable]:  # type: ignore
+
+        Returns:
+            An optional string representing the new album path configuration.
+            If None is returned, the default configuration will be used.
```
I think "default configuration" may imply Moe's default configuration, which isn't the case, but rather whatever is specified in the user's configuration file will be used.

jtpavlock commented on tests/move/test_move_core.py:
```diff
@@ -510,3 +514,114 @@ def test_move_core(self, tmp_config):
+            return "Classical/{album.artist}/{album.title} ({album.year})"
+        elif "Soundtrack" in album.title:
+            return "Soundtracks/{album.title} ({album.year})"
+        return None
```
Not sure if you mean to explicitly return None for clarity purposes, but it's not necessary in Python.

ali-ramadhan replied:
I do tend to `return None` for extra clarity, but I've removed it here.

jtpavlock reviewed (changes requested):
This looks great! I'm especially impressed with the tests, I know it probably wasn't easy trying to make sense of how I've structured my testing. I think just a few minor changes, and then clean up the commits and it should be good to go. 

For the commit, just a single feature commit would be perfect, no need to have a separate one for tests or anything.

ali-ramadhan commented on tests/move/test_move_core.py:
```diff
@@ -510,3 +514,114 @@ def test_move_core(self, tmp_config):
+        assert path == expected_path
+
+    @pytest.mark.usefixtures("_tmp_album_path_config")
+    def test_soundtrack_album(self):
```
Agreed! I've removed `test_soundtrack_album`.

ali-ramadhan commented on tests/move/test_move_core.py:
```diff
@@ -1,13 +1,17 @@
 """Tests the core api for moving items."""
 
+import datetime
```
Agreed and done!

ali-ramadhan commented on moe/move/move_core.py:
```diff
@@ -37,6 +37,22 @@ def create_path_template_func() -> list[Callable]:  # type: ignore
         """  # noqa: DAR202
 
+    @staticmethod
+    @moe.hookspec
```
Nice catch and done!

ali-ramadhan commented:
Thanks for the thorough review @jtpavlock and apologies for dropping the ball on this PR. I believe I've addressed your code review comments so I'm requesting another review.

I know the linter is complaining about stuff in `dup_core.py` even though this PR did not touch that file (and linting passes on the master branch). Happy to fix it here though.

> For the commit, just a single feature commit would be perfect, no need to have a separate one for tests or anything.

Yeah sorry I do tend to make lots of commits then squash and merge haha. What's the best solution here? I can still squash and merge to create a `feat(move):` commit. Or I can squash the commits I have made so far.

ali-ramadhan commented:
Alright, managed to squash all commits into a single `feat(move)` commit.

jtpavlock commented:
Thanks for coming back to this! I'll give it a review as soon as I can. You also should be able to rebase onto the newest main to get rid of the linting error. Every now and then a new pyright version will find some new issue with my typing.

jtpavlock reviewed (changes requested):
This looks awesome, thanks for the changes. It's nice, it ended up being a fairly simple code change to implement. The only thing left I realized we should probably do is mention or link to the hook in the docs in the path configuration section.

ali-ramadhan commented:
Agreed. I just added a note in the path configuration section mentioning the new hook!
